hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java

```diff
@@ -147,15 +160,33 @@ public void fire() {
 
             batchData.clear();
 
-            if (log.isTraceEnabled()) {
-                log.trace("Drools session dumping before firing: ");
+
+            TreeSet<Event> batchEvent = new TreeSet<Event>(pendingEvent);
+            Event previousEvent = null;
+
+            pendingEvent.clear();
+
+            for (Event event : batchEvent) {
+                if (null == previousEvent
+                        || (null != event.getDataId() && !event.getDataId().equals(previousEvent.getDataId()))) {
+                    kSession.insert(event);
+                    previousEvent = event;
+                } else {
+                    pendingEvent.add(event);
+                    log.debugf("Deferring more recent %1$s until older %2$s is processed", event, previousEvent);
+                }
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debugf("Firing cycle [%s] - with these facts: ", fireCycle);
                 for (FactHandle fact : kSession.getFactHandles()) {
                     Object o = kSession.getObject(fact);
-                    log.trace("Fact: " + o.toString());
+                    log.debugf("Fact:  %s", o.toString());
                 }
             }
 
             kSession.fireAllRules();
+            fireCycle++;
```

Comment:
Debugging is an ancient art :-) I thought you were serious. I'll leave it as int.